src/map/ai/controllers/pet_controller.cpp

```diff
@@ -40,10 +40,30 @@ void CPetController::Tick(time_point tick)
     TracyZoneScoped;
     TracyZoneString(PPet->getName());
 
-    if (PPet->objtype == TYPE_PET && static_cast<CPetEntity*>(PPet)->shouldDespawn(tick))
+    bool isPlayerPet = PPet->objtype == TYPE_PET || (PPet->objtype == TYPE_MOB && PPet->PMaster && PPet->PMaster->objtype == TYPE_PC);
+
+    // if a player pet then check if a charmed mob or jug pet and if it should despawn
+    if (isPlayerPet)
     {
-        petutils::DespawnPet(PPet->PMaster);
-        return;
+        // if a charmed mob and charm time is up then despawn
+        if (PPet->isCharmed && tick > PPet->charmTime)
+        {
+            petutils::DespawnPet(PPet->PMaster);
+            return;
+        }
+
+        // if a jug pet and the current time > jug spawn time + jug duration then despawn
+        auto* PPetEntity = dynamic_cast<CPetEntity*>(PPet);
+        if (PPetEntity && PPetEntity->isAlive() && PPetEntity->getPetType() == PET_TYPE::JUG_PET)
+        {
+            // need to covert tick to unix time (in seconds) because getJugSpawnTime and getJugDuration give unix time
+            auto tickAsUnixTime = static_cast<int32>(std::chrono::duration_cast<std::chrono::seconds>(tick.time_since_epoch()).count());
```

Comment:
Should this be uint32? A time can't be negative. If getJugSpawnTime and getJugDuration are returning int32 they need to be changed too. Also, if these are ints in seconds, are there any situations where the duration can be fractions of a second?

Reply:
> Should this be uint32? A time can't be negative. If getJugSpawnTime and getJugDuration are returning int32 they need to be changed too. Also, if these are ints in seconds, are there any situations where the duration can be fractions of a second?

I updated so that jugSpawnTime and jugDuration are uint32 everywhere they can be. The only place they cannot be is when they are saved as charVars (for persisting jug pets when zoning) as charVars are int32. The durations cannot be fractions of a second because they are read from the database via uint reader and then converted to a duration with a `seconds` construction. I am not sure if all the castings I do are fully needed but I guess better to be explicit rather than implicit. Thanks for the comments!